File: backend/schemas/plagiarism.py
"""
Schemas for Plagiarism Detection Feature
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict
from datetime import datetime
from enum import Enum


class SimilarityLevel(str, Enum):
    """Levels of similarity detection."""
    NONE = "none"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    VERY_HIGH = "very_high"


class MatchType(str, Enum):
    """Types of code matching."""
    EXACT = "exact"
    STRUCTURAL = "structural"  # AST structure match
    TOKEN_SEQUENCE = "token_sequence"  # Token pattern match
    RENAMED = "renamed"  # Same structure, different variable names
    PARTIAL = "partial"


class CodeMatch(BaseModel):
    """A matched code segment between two submissions."""
    match_type: MatchType = Field(..., description="Type of match detected")
    similarity: float = Field(..., ge=0, le=1, description="Similarity score (0-1)")
    code_snippet_1: str = Field("", description="Code snippet from first submission")
    code_snippet_2: str = Field("", description="Code snippet from second submission")
    line_range_1: tuple = Field((1, 1), description="Line range in first submission")
    line_range_2: tuple = Field((1, 1), description="Line range in second submission")
    explanation: str = Field("", description="Explanation of the match")


class SubmissionComparison(BaseModel):
    """Comparison result between two submissions."""
    submission_id_1: str = Field(..., description="First submission ID")
    submission_id_2: str = Field(..., description="Second submission ID")
    student_id_1: str = Field(..., description="First student ID")
    student_id_2: str = Field(..., description="Second student ID")
    similarity_score: float = Field(..., ge=0, le=1, description="Overall similarity (0-1)")
    matches: List[CodeMatch] = Field(default_factory=list, description="List of matching segments")
    analysis_notes: str = Field("", description="Analysis notes")


class PlagiarismCheckRequest(BaseModel):
    """Request for plagiarism check."""
    submission_id: str = Field(..., description="Submission ID")
    student_id: str = Field(..., description="Student ID of submission to check")
    course_id: str = Field(..., description="Course ID")
    code: str = Field(..., description="Code to check for plagiarism")
    compare_with_history: bool = Field(True, description="Compare with historical submissions")
    similarity_threshold: float = Field(0.7, ge=0, le=1, description="Threshold for flagging (0-1)")


class PlagiarismReport(BaseModel):
    """Comprehensive plagiarism detection report."""
    report_id: str = Field(..., description="Unique report identifier")
    submission_id: str = Field(..., description="Submission ID checked")
    checked_at: datetime = Field(..., description="Timestamp of check")

    # Overall Results
    overall_similarity: float = Field(..., ge=0, le=1, description="Highest similarity found (0-1)")
    similarity_level: SimilarityLevel = Field(..., description="Overall similarity level")
    is_flagged: bool = Field(..., description="Whether submission is flagged")

    # Detailed Comparisons
    comparisons: List[SubmissionComparison] = Field(default_factory=list)

    # Summary
    summary: str = Field("", description="Human-readable summary")


class SubmissionRecord(BaseModel):
    """Record of a submission for plagiarism database."""
    submission_id: str = Field(..., description="Unique submission ID")
    student_id: str = Field(..., description="Student ID")
    assignment_id: str = Field(..., description="Assignment ID")
    code: str = Field(..., description="Submitted code")
    code_hash: str = Field(..., description="Hash of normalized code")
    ast_fingerprint: str = Field(..., description="AST-based fingerprint")
    submitted_at: datetime = Field(..., description="Submission timestamp")
    token_sequence: List[str] = Field(default_factory=list, description="Tokenized code sequence")


class BatchPlagiarismRequest(BaseModel):
    """Request for batch plagiarism check."""
    assignment_id: str = Field(..., description="Assignment ID")
    submissions: List[PlagiarismCheckRequest] = Field(..., description="List of submissions to check")
    cross_compare: bool = Field(True, description="Compare submissions against each other")


class BatchPlagiarismReport(BaseModel):
    """Batch plagiarism check report."""
    report_id: str = Field(..., description="Unique report ID")
    course_id: str = Field(..., description="Course ID")
    checked_at: datetime = Field(..., description="Timestamp of check")
    total_submissions: int = Field(..., description="Total submissions checked")
    flagged_pairs: int = Field(..., description="Number of flagged pairs")
    comparisons: List[SubmissionComparison] = Field(default_factory=list)
    summary: str = Field("", description="Summary of batch check")


# ============================================
# 新增：增强的查重与原创性分析类型
# ============================================

class SimilarityAlgorithm(str, Enum):
    """相似度计算算法类型"""
    AST = "ast"                    # AST结构相似度
    LEVENSHTEIN = "levenshtein"    # 编辑距离算法
    COSINE = "cosine"              # 余弦相似度（TF-IDF）
    TOKEN = "token"                # Token序列相似度
    COMBINED = "combined"          # 综合算法


class CodeTransformationType(str, Enum):
    """代码变换类型"""
    VARIABLE_RENAME = "variable_rename"      # 变量重命名
    FUNCTION_RENAME = "function_rename"      # 函数重命名
    EXTRACT_FUNCTION = "extract_function"    # 提取函数
    INLINE_VARIABLE = "inline_variable"      # 内联变量
    REORDER_STATEMENTS = "reorder_statements"  # 语句重排序
    COMMENT_MODIFICATION = "comment_modification"  # 注释修改
    WHITESPACE_CHANGE = "whitespace_change"  # 空白符修改


class DetailedCodeMatch(BaseModel):
    """详细的代码匹配信息，包含精确位置"""
    source_student_id: str = Field(..., description="源学生ID")
    target_student_id: str = Field(..., description="目标学生ID")
    source_start_line: int = Field(..., ge=1, description="源代码起始行")
    source_end_line: int = Field(..., ge=1, description="源代码结束行")
    source_start_col: int = Field(0, ge=0, description="源代码起始列")
    source_end_col: int = Field(0, ge=0, description="源代码结束列")
    target_start_line: int = Field(..., ge=1, description="目标代码起始行")
    target_end_line: int = Field(..., ge=1, description="目标代码结束行")
    target_start_col: int = Field(0, ge=0, description="目标代码起始列")
    target_end_col: int = Field(0, ge=0, description="目标代码结束列")
    source_snippet: str = Field(..., description="源代码片段")
    target_snippet: str = Field(..., description="目标代码片段")
    similarity: float = Field(..., ge=0, le=1, description="相似度分数")
    algorithm: SimilarityAlgorithm = Field(..., description="检测算法")
    transformation_type: Optional[CodeTransformationType] = Field(
        None, description="检测到的代码变换类型"
    )
    explanation: str = Field("", description="匹配说明")


class SimilarityMatrixEntry(BaseModel):
    """相似度矩阵单元格数据"""
    student_id_1: str = Field(..., description="学生1 ID")
    student_id_2: str = Field(..., description="学生2 ID")
    student_name_1: str = Field("", description="学生1姓名")
    student_name_2: str = Field("", description="学生2姓名")
    similarity_score: float = Field(..., ge=0, le=1, description="综合相似度")
    algorithm_scores: Dict[str, float] = Field(
        default_factory=dict, description="各算法的相似度分数"
    )
    is_flagged: bool = Field(False, description="是否标记为可疑")


class SimilarityMatrix(BaseModel):
    """完整的相似度矩阵"""
    report_id: str = Field(..., description="报告ID")
    assignment_id: str = Field(..., description="作业ID")
    created_at: datetime = Field(..., description="创建时间")
    student_ids: List[str] = Field(..., description="学生ID列表")
    student_names: List[str] = Field(default_factory=list, description="学生姓名列表")
    matrix: List[List[float]] = Field(..., description="相似度矩阵数据")
    entries: List[SimilarityMatrixEntry] = Field(
        default_factory=list, description="详细的矩阵条目"
    )
    threshold: float = Field(0.7, description="相似度阈值")
    flagged_count: int = Field(0, description="标记的可疑对数量")


class OriginalityReport(BaseModel):
    """原创性分析报告"""
    report_id: str = Field(..., description="报告ID")
    submission_id: str = Field(..., description="提交ID")
    student_id: str = Field(..., description="学生ID")
    student_name: str = Field("", description="学生姓名")
    assignment_id: str = Field(..., description="作业ID")
    created_at: datetime = Field(..., description="创建时间")

    # 原创性评分
    originality_score: float = Field(..., ge=0, le=100, description="原创性评分(0-100)")

    # 各算法的相似度分解
    similarity_breakdown: Dict[str, float] = Field(
        default_factory=dict, description="各算法相似度分解"
    )

    # 详细匹配信息
    detailed_matches: List[DetailedCodeMatch] = Field(
        default_factory=list, description="详细的代码匹配列表"
    )

    # 相似的提交
    similar_submissions: List[str] = Field(
        default_factory=list, description="相似的提交ID列表"
    )

    # 检测到的代码变换
    detected_transformations: List[CodeTransformationType] = Field(
        default_factory=list, description="检测到的代码变换类型"
    )

    # 改进建议
    improvement_suggestions: List[str] = Field(
        default_factory=list, description="改进建议列表"
    )

    # 总结
    summary: str = Field("", description="分析总结")
    risk_level: SimilarityLevel = Field(
        SimilarityLevel.NONE, description="风险等级"
    )


class SubmissionData(BaseModel):
    """提交数据（用于批量分析）"""
    student_id: str = Field(..., description="学生ID")
    student_name: str = Field("", description="学生姓名")
    code: str = Field(..., description="代码内容")
    submission_id: Optional[str] = Field(None, description="提交ID")


class BatchAnalysisRequest(BaseModel):
    """批量分析请求"""
    assignment_id: str = Field(..., description="作业ID")
    course_id: str = Field("", description="课程ID")
    submissions: List[SubmissionData] = Field(..., description="提交列表")
    similarity_threshold: float = Field(
        0.7, ge=0, le=1, description="相似度阈值"
    )
    algorithms: List[SimilarityAlgorithm] = Field(
        default_factory=lambda: [SimilarityAlgorithm.COMBINED],
        description="使用的算法列表"
    )
    generate_reports: bool = Field(
        True, description="是否生成详细报告"
    )


class BatchAnalysisResponse(BaseModel):
    """批量分析响应"""
    report_id: str = Field(..., description="报告ID")
    assignment_id: str = Field(..., description="作业ID")
    created_at: datetime = Field(..., description="创建时间")
    total_submissions: int = Field(..., description="总提交数")
    total_comparisons: int = Field(..., description="总比较次数")
    flagged_count: int = Field(..., description="标记的可疑对数量")

    # 相似度矩阵
    similarity_matrix: SimilarityMatrix = Field(..., description="相似度矩阵")

    # 可疑对列表
    suspicious_pairs: List[SubmissionComparison] = Field(
        default_factory=list, description="可疑的提交对"
    )

    # 各学生的原创性报告
    originality_reports: List[OriginalityReport] = Field(
        default_factory=list, description="原创性报告列表"
    )

    # 总结
    summary: str = Field("", description="分析总结")


class PlagiarismSettings(BaseModel):
    """查重设置"""
    similarity_threshold: float = Field(
        0.7, ge=0, le=1, description="相似度阈值"
    )
    algorithms: List[SimilarityAlgorithm] = Field(
        default_factory=lambda: [SimilarityAlgorithm.COMBINED],
        description="启用的算法"
    )
    ast_weight: float = Field(0.3, ge=0, le=1, description="AST算法权重")
    token_weight: float = Field(0.25, ge=0, le=1, description="Token算法权重")
    text_weight: float = Field(0.2, ge=0, le=1, description="文本算法权重")
    semantic_weight: float = Field(0.15, ge=0, le=1, description="语义算法权重")
    order_invariant_weight: float = Field(0.1, ge=0, le=1, description="顺序不变算法权重")
    detect_renaming: bool = Field(True, description="检测变量/函数重命名")
    detect_refactoring: bool = Field(True, description="检测代码重构")

File: frontend-react-backup/src/types/plagiarism.ts
/**
 * 查重与原创性分析系统 - 类型定义
 */

// 相似度算法类型
export type SimilarityAlgorithm = 'ast' | 'levenshtein' | 'cosine' | 'token' | 'combined';

// 代码变换类型
export type CodeTransformationType =
  | 'variable_rename'
  | 'function_rename'
  | 'extract_function'
  | 'inline_variable'
  | 'reorder_statements'
  | 'comment_modification'
  | 'whitespace_change';

// 相似度等级
export type SimilarityLevel = 'none' | 'low' | 'medium' | 'high' | 'very_high';

// 匹配类型
export type MatchType = 'exact' | 'structural' | 'token_sequence' | 'renamed' | 'partial';

// 详细代码匹配信息
export interface DetailedCodeMatch {
  source_student_id: string;
  target_student_id: string;
  source_start_line: number;
  source_end_line: number;
  source_start_col: number;
  source_end_col: number;
  target_start_line: number;
  target_end_line: number;
  target_start_col: number;
  target_end_col: number;
  source_snippet: string;
  target_snippet: string;
  similarity: number;
  algorithm: SimilarityAlgorithm;
  transformation_type?: CodeTransformationType;
  explanation: string;
}

// 代码匹配
export interface CodeMatch {
  match_type: MatchType;
  similarity: number;
  code_snippet_1: string;
  code_snippet_2: string;
  line_range_1: [number, number];
  line_range_2: [number, number];
  explanation: string;
}

// 相似度矩阵条目
export interface SimilarityMatrixEntry {
  student_id_1: string;
  student_id_2: string;
  student_name_1: string;
  student_name_2: string;
  similarity_score: number;
  algorithm_scores: Record<string, number>;
  is_flagged: boolean;
}

// 相似度矩阵
export interface SimilarityMatrix {
  report_id: string;
  assignment_id: string;
  created_at: string;
  student_ids: string[];
  student_names: string[];
  matrix: number[][];
  entries: SimilarityMatrixEntry[];
  threshold: number;
  flagged_count: number;
}

// 提交对比结果
export interface SubmissionComparison {
  submission_id_1: string;
  submission_id_2: string;
  student_id_1: string;
  student_id_2: string;
  similarity_score: number;
  matches: CodeMatch[];
  analysis_notes: string;
}

// 原创性报告
export interface OriginalityReport {
  report_id: string;
  submission_id: string;
  student_id: string;
  student_name: string;
  assignment_id: string;
  created_at: string;
  originality_score: number;
  similarity_breakdown: Record<string, number>;
  detailed_matches: DetailedCodeMatch[];
  similar_submissions: string[];
  detected_transformations: CodeTransformationType[];
  improvement_suggestions: string[];
  summary: string;
  risk_level: SimilarityLevel;
}

// 提交数据
export interface SubmissionData {
  student_id: string;
  student_name?: string;
  code: string;
  submission_id?: string;
}

// 批量分析请求
export interface BatchAnalysisRequest {
  assignment_id: string;
  course_id?: string;
  submissions: SubmissionData[];
  similarity_threshold?: number;
  algorithms?: SimilarityAlgorithm[];
  generate_reports?: boolean;
}

// 批量分析响应
export interface BatchAnalysisResponse {
  report_id: string;
  assignment_id: string;
  created_at: string;
  total_submissions: number;
  total_comparisons: number;
  flagged_count: number;
  similarity_matrix: SimilarityMatrix;
  suspicious_pairs: SubmissionComparison[];
  originality_reports: OriginalityReport[];
  summary: string;
}

// 查重设置
export interface PlagiarismSettings {
  similarity_threshold: number;
  algorithms: SimilarityAlgorithm[];
  ast_weight: number;
  token_weight: number;
  text_weight: number;
  detect_renaming: boolean;
  detect_refactoring: boolean;
}

// 上传的文件信息
export interface UploadedFile {
  file: File;
  studentId: string;
  studentName: string;
  code: string;
  status: 'pending' | 'reading' | 'ready' | 'error';
  error?: string;
}

// 教师提交作业的请求类型
export interface TeacherAssignmentSubmit {
  assignment_id: string;
  course_id: string;
  title: string;
  description: string;
  due_date: string; // ISO 8601 format
  max_score?: number;
  assignment_type?: string;
  file_attachment?: ArrayBuffer; // 文件内容
  file_name?: string;
  rubric?: Record<string, any>;
  instructions?: string;
}

// 批量上传请求类型
export interface BatchUploadRequest {
  assignment_id: string;
  course_id: string;
  student_submissions: Array<{
    student_id: string;
    student_name: string;
    file_content: string;
    file_name: string;
  }>;
  sync_to_file_manager: boolean;
}

// 文件管理同步请求类型
export interface FileManagerSyncRequest {
  assignment_id: string;
  course_id: string;
  sync_type: 'full' | 'incremental';
  target_path: string;
  include_submissions?: boolean;
  include_feedback?: boolean;
}

// 文件管理同步响应类型
export interface FileManagerSyncResponse {
  sync_id: string;
  assignment_id: string;
  course_id: string;
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled';
  progress: number;
  total_files: number;
  processed_files: number;
  failed_files: number;
  start_time: string;
  end_time?: string;
  error_message?: string;
}

// 作业提交记录类型
export interface AssignmentSubmissionRecord {
  submission_id: string;
  assignment_id: string;
  student_id: string;
  student_name: string;
  file_name: string;
  file_size: number;
  submitted_at: string;
  status: string;
  content_preview: string;
}

// 作业传输响应类型
export interface AssignmentTransferResponse {
  transfer_id: string;
  assignment_id: string;
  course_id: string;
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled';
  total_submissions: number;
  successful_transfers: number;
  failed_transfers: number;
  start_time: string;
  end_time?: string;
  error_details?: Array<Record<string, any>>;
}

// 文件验证结果类型
export interface FileValidationResult {
  filename: string;
  is_valid: boolean;
  validation_errors: string[];
  file_size: number;
  detected_language?: string;
}

// 批量上传响应类型
export interface BatchUploadResponse {
  upload_id: string;
  assignment_id: string;
  course_id: string;
  total_files: number;
  successfully_uploaded: number;
  failed_uploads: number;
  validation_results: FileValidationResult[];
  start_time: string;
  end_time?: string;
}

